src/DotVVM.Framework/Hosting/DotvvmRequestContextExtensions.cs

```diff
@@ -83,14 +85,37 @@ public static void RedirectToUrl(this IDotvvmRequestContext context, string url,
         /// <summary>
         /// Returns the redirect response and interrupts the execution of current request.
         /// </summary>
-        public static void RedirectToRoute(this IDotvvmRequestContext context, string routeName, object newRouteValues = null, bool replaceInHistory = false, bool allowSpaRedirect = true, string urlSuffix = null)
+        public static void RedirectToRoute(this IDotvvmRequestContext context, string routeName, object newRouteValues = null, bool replaceInHistory = false, bool allowSpaRedirect = true, string urlSuffix = null, object queryStringParameters = null)
         {
             var route = context.Configuration.RouteTable[routeName];
-            var url = route.BuildUrl(context.Parameters, newRouteValues) + urlSuffix;
+            var url = route.BuildUrl(context.Parameters, newRouteValues) + BuildUrlSuffix(urlSuffix ?? "", queryStringParameters);
 
             context.RedirectToUrl(url, replaceInHistory, allowSpaRedirect);
         }
 
+        private static string BuildUrlSuffix(string urlSuffix, object queryStringParameters)
+        {
+            var hashIndex = urlSuffix.IndexOf('#');
+            var resultSuffix = hashIndex < 0 ? urlSuffix : urlSuffix.Substring(0, hashIndex);
+            if (queryStringParameters is IEnumerable<KeyValuePair<string, string>>)
+            {
+                foreach (KeyValuePair<string, string> item in (IEnumerable<KeyValuePair<string, string>>)queryStringParameters)
+                {
+                    resultSuffix = resultSuffix + (urlSuffix.LastIndexOf('?') < 0 ? "?" : "&") + Regex.Escape(item.Key) +
+                                "=" + Regex.Escape(item.Value);
```

Comment:
`Regex.Escape` IMO does not make any sense for escaping urls, did you wanted to use `Uri.EscapeDataString` instead?